pkg/kube/kube_handler.go

```diff
@@ -844,9 +863,22 @@ func filterActiveNamespaces(namespaces []corev1.Namespace) []corev1.Namespace {
 }
 
 // GetNamespaces return the list of namespaces that the user has permission to access
-func (a *userHandler) GetNamespaces() ([]corev1.Namespace, error) {
-	// Try to list namespaces with the user token, for backward compatibility
+func (a *userHandler) GetNamespaces(trustedNamespaces []corev1.Namespace) ([]corev1.Namespace, error) {
 	var namespaceList []corev1.Namespace
+
+	if len(trustedNamespaces) > 0 {
+		namespaceList = append(namespaceList, trustedNamespaces...)
+		// Filter namespaces that are in terminating state
+		namespaceList = filterActiveNamespaces(namespaceList)
+		// Filter namespaces in which the user has permissions to write (secrets) only
+		namespaceList, err := filterAllowedNamespaces(a.clientset, namespaceList)
+		if err != nil {
+			return nil, err
+		}
+		return namespaceList, nil
```

Comment:
Do we want to return empty list here or continue with the regular namespace check?

Reply:
I think the if/else condition here should only set the `namespaceList`:

```
if len(trustedNamespaces) > 0 {
    namespaceList = append(namespaceList, trustedNamespaces...)
} else {
    existing fetch from API as shown below
}

// Filter namespaces that are in terminating state
namespaceList = filterActiveNamespaces(namespaceList)
...
```

Also, it's not clear to me why you want to use `filterAllowedNamespaces` in the case where the namespaces have been provided as input? Can't we assume that the user can access these if we've been told so at this point? (`filterAllowedNamespaces` is only used when we know we've gotten a list that isn't restricted to those for the user only).

Reply:
> Also, it's not clear to me why you want to use `filterAllowedNamespaces` in the case where the namespaces have been provided as input? Can't we assume that the user can access these if we've been told so at this point? (`filterAllowedNamespaces` is only used when we know we've gotten a list that isn't restricted to those for the user only).

What happens if there is some drift though between the whitelisted namespaces and the ones the user can _actually_ access per k8s RBAC?

Reply:
Currently, we are completely relying on the namespaces returned by the header. Whereas it is supposed to be an admin configuration, what if the info returned by the API GW (headers) and does not match the one hold by the cluster?
Shouldn't we still perform the filtering based on the permissions?  

If not, what would be the user-facing expected error if that happens to occur?

Reply:
> Shouldn't we still perform the filtering based on the permissions?

Sorry, to be clear, by 'permissions' do you mean API gateway-specified permissions, or the k8s RBAC permissions?

Reply:
Sorry, yes, I meant the k8s RBAC permissions. Current implementation is not running `filterAllowedNamespaces` to check whether the user has permission to write secrets. So, in case the API gw gives a wrong namespace, we would be skipping the verification and wouldn't notice the error, no?

Reply:
I believe that's correct and in that case we'd need to now handle k8s API auth errors from the actually-not-whitelisted whitelisted namespaces... cc @goock

Reply:
> Sorry, yes, I meant the k8s RBAC permissions. Current implementation is not running `filterAllowedNamespaces` to check whether the user has permission to write secrets. So, in case the API gw gives a wrong namespace, we would be skipping the verification and wouldn't notice the error, no?

I had thought the intention was that if the headers are set, they are used as is and that it is an issue with the configuration if the gateway is including namespaces for which it should not (ie. no security issue, just the user *won't* be able to use them as RBAC will not allow it)? We could ensure that `filterAllowedNamespaces()` is also called for the defined namespaces, but can end up back in a similar position if many namespaces are configured.

If we assume that any configured namespaces from the header are correct, and they are *incorrectly* configured (ie. users able to select namespaces to which they cannot write), the error would be noticed when users select the namespace (if they don't have RBAC to read) or deploy to that namespace (if they can read, but can't write). On the other hand, if we do filter the namespaces provided by the header, it will hide the configuration error (users just won't have the namespace to select?)

IMO, the latter would be more of an issue (not detecting the configuration error where the gateway is adding namespaces for a user that it should not because we automatically filter them out), but don't feel strongly about it.

Reply:
>  We could ensure that filterAllowedNamespaces() is also called for the defined namespaces, but can end up back in a similar position if many namespaces are configured. 
> (...)  if we do filter the namespaces provided by the header, it will hide the configuration error (users just won't have the namespace to select?)

Mmm, I see. Yes, it makes sense just avoiding double checks if it is the gw which is performing this logic. As you said, if we filter them out, we will be hiding the configuration error, which is way worse than just showing an RBAC error.

Ok then, I just wanted to check it with you. Thanks for your comments; go ahead!